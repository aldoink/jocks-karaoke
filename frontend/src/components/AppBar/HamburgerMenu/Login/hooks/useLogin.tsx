import { useContext, useEffect, useState } from "react";
import { AuthContext } from "../../../../../contexts/AuthContext";

export enum LoginStatus {
  LOADING,
  SUCCESS,
  FAILURE,
  READY,
}

export function useLogin(email: string, password: string) {
  const [status, setStatus] = useState(LoginStatus.READY);
  const { authService, setIsAuthenticated } = useContext(AuthContext);

  useEffect(() => {
    setStatus(LoginStatus.READY);
  }, [email, password]);

  const login = async () => {
    setStatus(LoginStatus.LOADING);
    try {
      await authService.login(email, password);
      setStatus(LoginStatus.SUCCESS);
      setIsAuthenticated(true);
    } catch (error) {
      setStatus(LoginStatus.FAILURE);
    }
  };

  return { status, login };
}
