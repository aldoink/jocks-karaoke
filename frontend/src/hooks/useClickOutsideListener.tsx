import {RefObject, useEffect} from "react";

export const useClickOutsideListener = (ref: RefObject<HTMLElement>, callback: Function) => {
    useEffect(() => {
        const handleClickOutside = (event: any) => {
            if (!ref.current || ref.current.contains(event.target)) {
                return;
            }
            callback(event);
        }
        // Bind the event listener
        document.addEventListener("mousedown", handleClickOutside);
        document.addEventListener("touchstart", handleClickOutside);
        return () => {
            // Unbind the event listener on clean up
            document.removeEventListener("mousedown", handleClickOutside);
            document.removeEventListener("touchend", handleClickOutside);
        };
    }, [ref, callback]);
}