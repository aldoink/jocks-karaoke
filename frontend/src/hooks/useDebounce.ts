import {useEffect, useState} from "react";

export default function useDebounce(value: any, delay: number) {
    const [debouncedValue, setDebouncedValue] = useState(undefined);
    useEffect(() => {
            const handler = setTimeout(() => setDebouncedValue(value), delay);
            return () => clearTimeout(handler);
        },
        [value, delay]
    );
    return debouncedValue;
}