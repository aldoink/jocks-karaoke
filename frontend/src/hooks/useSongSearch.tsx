import {useContext, useEffect, useState} from "react";
import {SongContext} from "../contexts/SongContext";
import useDebounce from "./useDebounce";

export const useSongSearch = () => {
    const {setSongs, songService} = useContext(SongContext);
    const [searchTerm, setSearchTerm] = useState<string | undefined>(undefined);
    const debouncedSearchTerm: string | undefined = useDebounce(searchTerm, 300);

    useEffect(() => {
        async function callSongService() {
            if (debouncedSearchTerm || debouncedSearchTerm === '') {
                try {
                    const songList = await songService.search(debouncedSearchTerm);
                    setSongs(songList);
                } catch (e) {
                    //DO NOTHING
                }
            }
        }

        callSongService();
    }, [debouncedSearchTerm, setSongs, songService]);

    const search = (searchTerm: string) => setSearchTerm(searchTerm);

    const searchWithoutDebounce = async (searchTerm: string) => {
        try {
            const songList = await songService.search(searchTerm);
            setSongs(songList);
        } catch (e) {
            //DO NOTHING
        }
    }

    return {search, searchWithoutDebounce}
}