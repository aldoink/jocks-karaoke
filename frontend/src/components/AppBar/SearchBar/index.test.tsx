import {render, screen} from "@testing-library/react";
import {SearchBar} from "./index";
import {useSongSearch} from "../../../hooks/useSongSearch";
import userEvent from "@testing-library/user-event";

jest.mock("../../../hooks/useSongSearch");

describe('SearchBar', () => {

    const mockedSearchFn = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        (useSongSearch as any).mockReturnValue(mockedSearchFn);
    });

    it('renders correctly', () => {
        expect(render(<SearchBar/>).asFragment()).toMatchSnapshot();
    });

    it('calls the useSongSearch hook search function when user types in input', () => {
        //given
        render(<SearchBar/>);

        //when
        userEvent.type(screen.getByPlaceholderText('Search'), 'test');

        //then
        expect(mockedSearchFn).toHaveBeenCalledWith('t');
        expect(mockedSearchFn).toHaveBeenCalledWith('te');
        expect(mockedSearchFn).toHaveBeenCalledWith('tes');
        expect(mockedSearchFn).toHaveBeenCalledWith('test');
    });

    it('loads all songs on first load', () => {
        //given
        const mockedSearchFn = jest.fn();
        (useSongSearch as any).mockReturnValue(mockedSearchFn);
        render(<SearchBar/>);

        //then
        expect(mockedSearchFn).toHaveBeenCalledTimes(1);
        expect(mockedSearchFn).toHaveBeenCalledWith('');
    });
});