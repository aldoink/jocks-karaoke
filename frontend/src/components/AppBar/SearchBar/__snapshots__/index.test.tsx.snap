// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders correctly 1`] = `
<DocumentFragment>
  .c1 {
  padding: 0.5rem;
  background-color: transparent;
  border: none;
  border-bottom: 1px solid white;
  color: white;
  font-size: 16px;
  caret-color: white;
}

.c1::-webkit-outer-spin-button,
.c1::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c1[type=number] {
  -moz-appearance: textfield;
}

.c1:focus {
  outline: none;
}

.c1::-webkit-input-placeholder {
  color: white;
  opacity: 1;
}

.c1::-moz-placeholder {
  color: white;
  opacity: 1;
}

.c1:-ms-input-placeholder {
  color: white;
  opacity: 1;
}

.c1::placeholder {
  color: white;
  opacity: 1;
}

.c0 {
  position: relative;
  display: inline-block;
  -webkit-transition: 0.5s;
  transition: 0.5s;
  width: 100%;
  padding: 1rem;
}

.c0 svg {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  fill: white;
  height: 20px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  position: absolute;
  right: 1rem;
  bottom: 1.5rem;
}

.c2 {
  box-sizing: border-box;
  width: 100%;
  -webkit-transition: border-bottom-color 0.5s;
  transition: border-bottom-color 0.5s;
}

.c2:focus {
  border-bottom-color: red;
}

<div
    class="c0"
  >
    <input
      class="c1 c2"
      placeholder="Search"
      type="text"
    />
    <svg>
      search-icon.svg
    </svg>
  </div>
</DocumentFragment>
`;
